pkg/ignition/enrich.go

```diff
@@ -0,0 +1,157 @@
+type Enricher struct {
+	BootstrapData []byte
+	Hostname      string
+	InstanceID    string
+	ProviderID    string
+	Network       []cloudinit.NetworkConfigData
+}
+
+// Enrich enriches the Ignition config with additional data.
+func (e *Enricher) Enrich() ([]byte, string, error) {
+	return buildIgnitionConfig(e.BootstrapData, e.getEnrichConfig())
+}
+
+func (e *Enricher) getEnrichConfig() *ignitionTypes.Config {
+	ign := &ignitionTypes.Config{
+		Storage: ignitionTypes.Storage{
+			Files: []ignitionTypes.File{
+				{
+					Node: ignitionTypes.Node{
+						Filesystem: "root",
+						Path:       "/etc/hostname",
+						Overwrite:  ptr.To(true),
+					},
+					FileEmbedded1: ignitionTypes.FileEmbedded1{
+						Mode: ptr.To(0644),
+						Contents: ignitionTypes.FileContents{
+							Source: fmt.Sprintf("data:,%s", e.Hostname),
+						},
+					},
+				},
+				{
+					Node: ignitionTypes.Node{
+						Filesystem: "root",
+						Path:       "/etc/proxmox-env",
+						Overwrite:  ptr.To(true),
+					},
+					FileEmbedded1: ignitionTypes.FileEmbedded1{
+						Mode: ptr.To(420),
+						Contents: ignitionTypes.FileContents{
+							Source: fmt.Sprintf("data:,%s", e.getProxmoxEnvContent()),
+						},
+					},
+				},
+			},
+		},
+		Networkd: ignitionTypes.Networkd{
+			Units: []ignitionTypes.Networkdunit{},
+		},
+		Systemd: ignitionTypes.Systemd{
+			Units: []ignitionTypes.Unit{
+				{
+					Name:   "systemd-resolved.service",
+					Enable: true,
+				},
+			},
+		},
+	}
+
+	for i, net := range e.Network {
+		ign.Networkd.Units = append(ign.Networkd.Units, ignitionTypes.Networkdunit{
+			Name:     fmt.Sprintf("%d%d-eth%d.network", i, i, i),
+			Contents: getNetworkdUnit(net),
+		})
+	}
+
+	return ign
+}
+
+func getNetworkdUnit(net cloudinit.NetworkConfigData) string {
+	str := fmt.Sprintf("[Match]\nMACAddress=%s\n\n[Link]\nName=%s\n\n[Network]\nAddress=%s\nGateway=%s\n", net.MacAddress, net.Name, net.IPAddress, net.Gateway)
+	if net.IPV6Address != "" {
+		str += fmt.Sprintf("Address=%s\n", net.IPV6Address)
+		str += fmt.Sprintf("Gateway=%s\n", net.Gateway6)
+	}
+	if net.DNSServers != nil {
+		for _, dns := range net.DNSServers {
+			str += fmt.Sprintf("DNS=%s\n", dns)
+		}
+	}
+	return str
+}
+
+func (e *Enricher) getProxmoxEnvContent() string {
+	content := fmt.Sprintf("COREOS_CUSTOM_HOSTNAME=%s\nCOREOS_CUSTOM_INSTANCE_ID=%s\nCOREOS_CUSTOM_PROVIDER_ID=%s", e.Hostname, e.InstanceID, e.ProviderID)
+	if len(e.Network) > 0 && e.Network[0].IPAddress != "" {
+		content += fmt.Sprintf("\nCOREOS_CUSTOM_PRIVATE_IPV4=%s", e.Network[0].IPAddress)
+	}
+	if len(e.Network) > 0 && e.Network[0].IPV6Address != "" {
+		content += fmt.Sprintf("\nCOREOS_CUSTOM_PRIVATE_IPV6=%s", e.Network[0].IPV6Address)
+	}
+	return url.PathEscape(content)
+}
+
+func buildIgnitionConfig(bootstrapData []byte, enrichConfig *ignitionTypes.Config) ([]byte, string, error) {
+	// We control bootstrapData config, so treat it as strict.
```

Comment:
What exactly is meant by strict in this context?

Reply:
that means even parsing warnings are not accepted

Reply:
This feels like a useless comment in its current state. Either it should explain what strict means in more detail or just be removed because this comment will not help anyone debugging this code. At the moment you'll always have to look into `convertToIgnition` to understand what strict means.